segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,384 @@
+    @Override
+    public CompletableFuture<Integer> read(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int
+            length, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncRead(handle, offset, buffer, bufferOffset, length));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> getStreamSegmentInfo(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncGetStreamSegmentInfo(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncExists(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync(streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncWrite(handle, offset, data, length));
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+        return supplyAsync(targetHandle.getSegmentName(),
+                () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region private sync implementation
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openRead", streamSegmentName);
+
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtendedS3SegmentHandle retHandle = ExtendedS3SegmentHandle.getReadHandle(streamSegmentName);
+        LoggerHelpers.traceLeave(log, "openRead", traceId, streamSegmentName);
+        return retHandle;
+    }
+
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "openWrite", streamSegmentName);
+        StreamSegmentInformation info = syncGetStreamSegmentInfo(streamSegmentName);
+        ExtendedS3SegmentHandle retHandle;
+        if (info.isSealed()) {
+            retHandle = ExtendedS3SegmentHandle.getReadHandle(streamSegmentName);
```

Comment:
The contract doesn't say that if the segment is sealed, we will return a read handle:

```
@return A CompletableFuture that, when completed, will contain a read-write SegmentHandle that can be used to access
     * the segment for read and write activities (ex: read, get, write, seal, concat). If the operation failed, it will be
     * failed with the cause of the failure. Notable exceptions:
     * <ul>
     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
     * </ul>
```

There is an inconsistency either in this code or in the `Storage` contract.